### PR TITLE
Server-side criterium rendering.

### LINKED ISSUE #2029: Buyers Guide: audit the model for and `addslashes` need
Give https://github.com/mozilla/foundation.mozilla.org/pull/2010, which added `addslashes`, and then https://github.com/mozilla/foundation.mozilla.org/pull/2026, which selectively takes them away depending on a property name, we should audit the code to see:

1.  why we need `addslashes` at all (can we be better with the code we're generating?), and
2. whether we can cleanly set up something that'll yield the correct HTML without using nested template logic.

### LINKED ISSUE #2030: Inconsistency between prod and local dev: bugs on deletes_data/child_rules
/cc @cadecairos @alanmoo @kristinashu 

### what's wrong?

https://foundation.mofostaging.net/privacynotincluded/products/apple-watch-4/ shows off a bug in product page rendering, where its model looks like this:

image

but its associated product page looks like this:

image

Where React has created the following DOM (note the `criterion-undefined` entries):

image

Off of the following Django HTML:

```
<div class="mb-5">
 <div class="criterion-target" data-meta='{"class": "null","id": "must_change_default_password","question": "If a password is required, you must change the default password","helptext": "","answer": "Not required"}'></div>
 <div class="criterion-target" data-meta='{"class": "true","id": "security_updates","question": "Automatic security updates","helptext": "","answer": "Yes"}'></div>
 <div class="criterion-target" data-meta='{"class": "true","id": "delete_data","question": "Deletes data it stores on you","helptext": "","answer": "Yes"}'></div>
 <div class="criterion-target" data-meta='{"class": "null","id": "child_rules","question": "Has parental controls","helptext": "","answer": "Can\&#39;t determine"}'></div>
 </div>
```

### trying this locally

Copying the production database and loading it locally, I see the same model content:

image

But no errors on the associated product page:

image

Because React generated this DOM:

image

off of the following Django HTML:

```
<div class="mb-5">
 <div class="criterion-target" data-meta='{"class": "null","id": "must_change_default_password","question": "If a password is required, you must change the default password","helptext": "","answer": "Not required"}'></div>
 <div class="criterion-target" data-meta='{"class": "true","id": "security_updates","question": "Automatic security updates","helptext": "","answer": "Yes"}'></div>
 <div class="criterion-target" data-meta='{"class": "false","id": "delete_data","question": "Deletes data it stores on you","helptext": "","answer": "No"}'></div>
 <div class="criterion-target" data-meta='{"class": "true","id": "child_rules","question": "Has parental controls","helptext": "","answer": "Yes"}'></div>
 </div>
```

So the same codebase, on the same database, on two different computers, yielding two very different HTML sources. I ... don't know how this is possible.

(yet?)

### LINKED ISSUE #2031: Buyers Guide QA: if Reading level Can't be determined then it should say that
Example: https://foundation.mozilla.org/privacynotincluded/products/fredi-baby-monitor/ and https://foundation.mozilla.org/privacynotincluded/products/tractive-gps-3g-pet-tracker/

If the Reading Level in the CMS is set to "Can't Determine" then the page should say "Can't determine". But on Prod is says "Level 0" (this is a new issue, used to be ok).

CMS
screen shot 2018-10-23 at 5 02 18 pm

Prod
screen shot 2018-10-23 at 5 02 45 pm

### PR DESCRIPTION
**Pull request by Pomax**
Supercededs #2032

Closes #2029, closes #2031

### CODE REVIEW

File: source/sass/global.scss [hunk 1]
```diff
@@ -91,3 +91,7 @@ button:hover {
     color: transparentize($white, 0.5);
   }
 }
+
+.hidden {
```
Comment (alanmoo) [inline source/sass/global.scss]: Bootstrap should have a class for this already

Comment (Pomax) [reply source/sass/global.scss]: it does, `d-none`, which is a terrible name. I'll use it, but let the record show that the only right name for this CSS is `hidden`.

File: source/js/buyers-guide/components/criterion/criterion.jsx [hunk 2]
```diff
@@ -1,12 +1,37 @@
 import React from 'react';
 
+// Helper function to get the text content from
+// a DOM element. Returns false if the query
+// selector fails to find an element.
+function getValueFrom(e, qs) {
+  let v = e.querySelector(qs);
+  return v ? v.textContent : false;
```
Comment (Pomax) [inline source/js/buyers-guide/components/criterion/criterion.jsx]: not quite right yet, we want that HTML to actually do things.

File: network-api/networkapi/buyersguide/templates/fragments/product-criterion.html [hunk 3]
```diff
@@ -0,0 +1,20 @@
+<div class="criterion criterion-{{ value|lower }} criterion-{% if value == bad %}bad{% else %}good{% endif %}">
```
Comment (alanmoo) [inline network-api/networkapi/buyersguide/templates/fragments/product-criterion.html]: This seems like you're just outputting `criterion-bad criterion-bad` or `criterion-good criterion-good`. The existing template outputs the criterion name?

Comment (alanmoo) [reply network-api/networkapi/buyersguide/templates/fragments/product-criterion.html]: Ah, I think I see now. This has to do with the bad attribute being passed in to the fragment?

File: source/sass/buyers-guide/views/product.scss [hunk 4]
```diff
@@ -107,20 +107,7 @@
...
-.criterion-privacy_policy_reading_level.criterion-17,
-.criterion-privacy_policy_reading_level.criterion-18,
-.criterion-privacy_policy_reading_level.criterion-19,
-.criterion-share_data.criterion-true,
-.criterion-must_change_default_password.criterion-false,
-.criterion-security_updates.criterion-false,
-.criterion-delete_data.criterion-false,
-.criterion-child_rules.criterion-false,
-.criterion-manage_security.criterion-false {
+.criterion-bad {
```
Comment (alanmoo) [inline source/sass/buyers-guide/views/product.scss]: Looks like the mapping of "reading levels >=13 are bad" didn't get translated into the refactor.

**Comment by alanmoo** [comment]
@kristinashu I combed through this refactor/improvement, but could probably use a second set of eyes to make sure I'm not missing any functionality. Basically the "criteria" items have been refactored-  I edited [this item](https://foundation-mofostaging-pr-2036.herokuapp.com/privacynotincluded/products/wife-computer/) to set every item to Yes, No, and then Unknown, comparing it to [this staging item](https://foundation.mofostaging.net/privacynotincluded/products/petchatz-hd/). Yes and No looked fine, but "Unknown" items have different words and the reading levels need "bad" styling. If there's any other functionality I'm not aware of to check for, please note it here?

**Pull request merged**